components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1201 @@
+    CGPoint destination = [self placeholderFloatingPosition];
+    NSLayoutConstraint *top = [NSLayoutConstraint constraintWithItem:self.textInput.placeholderLabel
+                                                           attribute:NSLayoutAttributeTop
+                                                           relatedBy:NSLayoutRelationEqual
+                                                              toItem:self.textInput
+                                                           attribute:NSLayoutAttributeTop
+                                                          multiplier:1
+                                                            constant:destination.y];
+
+    NSLayoutConstraint *leading =
+        [NSLayoutConstraint constraintWithItem:self.textInput.placeholderLabel
+                                     attribute:NSLayoutAttributeLeading
+                                     relatedBy:NSLayoutRelationEqual
+                                        toItem:self.textInput
+                                     attribute:NSLayoutAttributeLeading
+                                    multiplier:1
+                                      constant:destination.x];
+    self.placeholderAnimationConstraints = @[ top, leading ];
+
+    self.previousPlaceholderColor = self.textInput.placeholderLabel.textColor;
+
+    animationBlock = ^{
+      self.textInput.placeholderLabel.transform = floatingPlaceholderScaleTransform;
+
+      self.textInput.placeholderLabel.textColor = self.textInput.tintColor;
+      [NSLayoutConstraint activateConstraints:self.placeholderAnimationConstraints];
+    };
+  } else {
+    animationBlock = ^{
+      self.textInput.placeholderLabel.transform = CGAffineTransformIdentity;
+
+      self.textInput.placeholderLabel.textColor =
+          self.previousPlaceholderColor ?: self.textInput.placeholderLabel.textColor;
+      [NSLayoutConstraint deactivateConstraints:self.placeholderAnimationConstraints];
+    };
+  }
+
+  // We do this beforehand to flush the layout engine.
+  [self.textInput layoutIfNeeded];
+  [UIView animateWithDuration:[CATransaction animationDuration]
+      animations:^{
+        animationBlock();
+        [self.textInput layoutIfNeeded];
+      }
+      completion:^(BOOL finished) {
+        if (!isToUp) {
+          self.placeholderAnimationConstraints = nil;
+        }
+      }];
+}
+
+- (CGPoint)placeholderFloatingPosition {
+  CGFloat placeholderY = MDCTextInputVerticalPadding;
+
+  // Offsets needed due to transform working on normal (0.5,0.5) anchor point.
+  // Why no anchor point of (0,0)? Because our users wouldn't expect it.
+  placeholderY -=
+      self.textInput.placeholderLabel.font.lineHeight * (1 - [self effectiveFloatingScale]) * .5f;
+
+  CGFloat estimatedWidth = MDCCeil(CGRectGetWidth([self.textInput.placeholderLabel.text
+      boundingRectWithSize:CGSizeMake(CGFLOAT_MAX,
+                                      MDCCeil(self.textInput.placeholderLabel.font.lineHeight))
```

Comment:
I was being super cautious but I removed it and ran it. I don't see any difference. Done.